lib/at_cmd_parser/at_cmd_parser.c

```diff
@@ -367,6 +367,91 @@ static int at_parse_param(const char **at_params_str,
 	return 0;
 }
 
+/*
+ * Internal function.
+ * Check AT command grammar based on below.
+ *  AT<terminator>
+ *  AT<separator><body><terminator>
+ *  AT<separator><body>=<terminator>
+ *  AT<separator><body>?<terminator>
+ *  AT<separator><body>=?<terminator>
+ *  AT<separator><body>=<parameters><terminator>
+ * In which
+ * <terminator>: CR, LF, CRLF, NULL
+ * <separator>: +, %, #
+ * <body>: alphabetic char only, size > 0
+ * <parameters>: arbitrary, size > 0
+ */
+int command_grammar_check(const char *at_params_str)
+{
+	const char *cmd = at_params_str;
+	const char *body = NULL;
+
+	/* check AT (if not, no check) */
+	if (strlen(cmd) < 2 || toupper((int)cmd[0]) != 'A' || toupper((int)cmd[1]) != 'T') {
+		return 0;
+	}
+	/* check AT<terminator> */
+	cmd += 2;
+	if (is_lfcr(*cmd) || is_terminated(*cmd)) {
+		return 0;
+	}
+	/* check AT<separator> */
+	if ((*cmd != AT_STANDARD_NOTIFICATION_PREFIX) && (*cmd != AT_PROP_NOTIFICATION_PREFX) &&
+	    (*cmd != AT_CUSTOM_COMMAND_PREFX)) {
+		return -EINVAL;
+	}
+	/* check AT<separator><body> */
+	cmd += 1;
```

Comment:
Actually, this whole part could be replaced with an already existing `is_command()` helper.

Reply:
It's a bit difficult. For example,
aat: is_command() returns false and here I should return 0;  /*  treated as AT response or unsolicited */
att:  is_command() returns false and here I should return -EINVAL;

Reply:
ok